clickhouse wrong password error is misleading

### Describe the bug

connect to clickhouse but use a wrong password

```docker
docker run -p8123:8123 -p9000:9000 -e"CLICKHOUSE_PASSWORD=password" -it --rm clickhouse/clickhouse-server:latest
```

Connect but use an incorrect password:

<img width="666" alt="Image" src="https://github.com/user-attachments/assets/01a0ffa7-1800-4123-aedf-d3ab80634aea" />

<details>
<summary> heft stacktrace but with clearly identifying info:</summary>

```
2025-06-27 21:22:41,532 ERROR driver.util :: Failed to connect to Database
java.sql.SQLException: Failed to create connection
	at com.clickhouse.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:104)
	at com.clickhouse.jdbc.Driver.connect(Driver.java:123)
	at com.clickhouse.jdbc.ClickHouseDriver.connect(ClickHouseDriver.java:97)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:683)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:191)
	at clojure.java.jdbc$get_driver_connection.invokeStatic(jdbc.clj:271)
	at clojure.java.jdbc$get_driver_connection.invoke(jdbc.clj:250)
	at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:411)
	at clojure.java.jdbc$get_connection.invoke(jdbc.clj:274)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1111)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
va.jdbc$query.invokeStatic(jdbc.clj:1160)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invokeStatic(connection.clj:368)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invoke(connection.clj:365)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_$fn__137759.invoke(connection.clj:377)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection$fn__137737.invoke(connection.clj:353)
	at metabase.driver.sql_jdbc.connection.ssh_tunnel$do_with_ssh_tunnel.invokeStatic(ssh_tunnel.clj:152)
	at metabase.driver.sql_jdbc.connection.ssh_tunnel$do_with_ssh_tunnel.invoke(ssh_tunnel.clj:141)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invokeStatic(connection.clj:348)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invoke(connection.clj:344)
tic(connection.clj:376)ver.sql_jdbc.connection$can_connect_QMARK_.invokeSta
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invoke(connection.clj:372)
	at metabase.driver.clickhouse$eval200658$fn__200659.invoke(clickhouse.clj:172)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__89713.invoke(util.clj:135)
	at clojure.core$binding_conveyor_fn$fn__5844.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: com.clickhouse.client.api.ClientException: Failed to get server info
	at com.clickhouse.client.api.Client.loadServerInfo(Client.java:204)
	at com.clickhouse.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:93)
	... 33 more
nt.api.ServerException: Code: 194. DB::Exception: default: Authentication failed: password is incorrect, or there is no user with such name. If you use ClickHouse Cloud, the password can be reset at https://clickhouse.cloud/ on the settings page for the corresponding service. If you have installed ClickHouse and forgot password you can reset it in the configuration file. The password for default user is typically located at /etc/clickhouse-server/users.d/default-password.xml and deleting this file will reset the password. See also /etc/clickhouse-server/users.xml on the server where ClickHouse is installed. . (REQUIRED_PASSWORD) (version 25.1.3.23 (official build))
	at com.clickhouse.client.api.internal.HttpAPIClientHelper.readError(HttpAPIClientHelper.java:371)
	at com.clickhouse.client.api.internal.HttpAPIClientHelper.executeRequest(HttpAPIClientHelper.java:426)
	at com.clickhouse.client.api.Client.lambda$query$10(Client.java:1705)
)	at com.clickhouse.client.api.Client.runAsyncOperation(Client.java:2134
	at com.clickhouse.client.api.Client.query(Client.java:1748)
	at com.clickhouse.client.api.Client.query(Client.java:1634)
	at com.clickhouse.client.api.Client.loadServerInfo(Client.java:195)
	... 34 more
2025-06-27 21:22:41,578 ERROR warehouses.api :: Cannot connect to Database
clojure.lang.ExceptionInfo: Failed to create connection {:message "Failed to create connection"}
	at metabase.driver.util$can_connect_with_details_QMARK_.invokeStatic(util.clj:152)
	at metabase.driver.util$can_connect_with_details_QMARK_.doInvoke(util.clj:123)
	at clojure.lang.RestFn.invoke(RestFn.java:445)
	at metabase.warehouses.api$test_database_connection.invokeStatic(api.clj:754)
	at metabase.warehouses.api$test_database_connection.doInvoke(api.clj:744)
	at clojure.lang.RestFn.invoke(RestFn.java:428)
	at metabase.warehouses.api$eval186382$fn__186389.invoke(api.clj:932)
	at metabase.api.macros$fn__149940$_core_fn149939__149941$core_fn__149942.invoke(macros.clj:415)
	at metabase.api.macros$fn__150020$endpoint_handler_STAR_150019__150022$handler__150026.invoke(macros.clj:611)
	at metabase.api.macros$fn__150073$build_ns_handler150072__150074$ns_handler_STAR___150075.invoke(macros.clj:675)
WithOpenAPISpec.invoke(open_api.clj:42)r
	at metabase.api.macros$fn__150120$ns_handler150119__150121$dev_handler__150126.invoke(macros.clj:774)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.routes.common$enforce_authentication$fn__173491.invoke(common.clj:91)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.util.handlers$_route_map_handler$fn__151968.invoke(handlers.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at compojure.core$routes$fn__151809$f__151810.invoke(core.clj:198)
	at compojure.core$routes$fn__151809$f__151810$respond_SINGLEQUOTE___151811.invoke(core.clj:197)
	at metabase.api.util.handlers$_route_map_handler$fn__151968.invoke(handlers.clj:24)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at clojure.lang.Var.invoke(Var.java:395)
	at compojure.core$routes$fn__151809$f__151810.invoke(core.clj:198)
core.clj:200)mpojure.core$routes$fn__151809.invoke(
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at clojure.lang.Var.invoke(Var.java:395)
	at metabase.server.routes$fn__155523$api_handler155522__155524$api_handler_STAR___155525.invoke(routes.clj:71)
	at compojure.core$routes$fn__151809$f__151810.invoke(core.clj:198)
	at compojure.core$routes$fn__151809.invoke(core.clj:200)
	at compojure.core$make_context$handler__151877.invoke(core.clj:290)
	at compojure.core$make_context$fn__151881.invoke(core.clj:300)
	at compojure.core$routes$fn__151809$f__151810.invoke(core.clj:198)
	at compojure.core$routes$fn__151809$f__151810$respond_SINGLEQUOTE___151811.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__151790.invoke(core.clj:153)
	at compojure.core$routes$fn__151809$f__151810.invoke(core.clj:198)
	at compojure.core$routes$fn__151809$f__151810$respond_SINGLEQUOTE___151811.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__151790.invoke(core.clj:153)
151810.invoke(core.clj:198)outes$fn__151809$f__
	at compojure.core$routes$fn__151809$f__151810$respond_SINGLEQUOTE___151811.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__151790.invoke(core.clj:153)
	at compojure.core$routes$fn__151809$f__151810.invoke(core.clj:198)
	at compojure.core$routes$fn__151809$f__151810$respond_SINGLEQUOTE___151811.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__151790.invoke(core.clj:153)
	at compojure.core$routes$fn__151809$f__151810.invoke(core.clj:198)
	at compojure.core$routes$fn__151809$f__151810$respond_SINGLEQUOTE___151811.invoke(core.clj:197)
	at metabase.api.util.handlers$_route_map_handler$fn__151968.invoke(handlers.clj:24)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.util.handlers$_route_map_handler$fn__151968.invoke(handlers.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at clojure.lang.Var.invoke(Var.java:395)
.clj:198)t compojure.core$routes$fn__151809$f__151810.invoke(core
	at compojure.core$routes$fn__151809.invoke(core.clj:200)
	at metabase.server.middleware.exceptions$catch_uncaught_exceptions$fn__150597.invoke(exceptions.clj:97)
	at metabase.server.middleware.exceptions$catch_api_exceptions$fn__150594.invoke(exceptions.clj:86)
	at metabase.server.middleware.log$log_api_call$fn__151039$fn__151040$fn__151041.invoke(log.clj:224)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invokeStatic(diagnostic.clj:17)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invoke(diagnostic.clj:12)
	at metabase.server.middleware.log$log_api_call$fn__151039$fn__151040.invoke(log.clj:215)
	at toucan2.execute$do_with_call_counts.invokeStatic(execute.clj:112)
	at toucan2.execute$do_with_call_counts.invoke(execute.clj:103)
	at metabase.server.middleware.log$log_api_call$fn__151039.invoke(log.clj:214)
	at metabase.server.middleware.browser_cookie$ensure_browser_id_cookie$fn__150234.invoke(browser_cookie.clj:40)
iddleware.security$add_security_headers$fn__150553.invoke(security.clj:277)
	at metabase.server.middleware.json$wrap_json_body$fn__150669.invoke(json.clj:128)
	at metabase.server.middleware.offset_paging$handle_paging$fn__151130.invoke(offset_paging.clj:48)
	at metabase.server.middleware.json$wrap_streamed_json_response$fn__150657.invoke(json.clj:83)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__151465.invoke(keyword_params.clj:55)
	at ring.middleware.params$wrap_params$fn__151492.invoke(params.clj:77)
	at metabase.server.middleware.misc$maybe_set_site_url$fn__151087.invoke(misc.clj:59)
	at metabase.server.middleware.session$reset_session_timeout$fn__148187.invoke(session.clj:270)
	at metabase.server.middleware.session$bind_current_user$fn__148180$fn__148181.invoke(session.clj:245)
	at metabase.request.session$do_with_current_user$fn__110526.invoke(session.clj:51)
	at metabase.request.session$do_with_current_user$do_with_user_local_values__110521.invoke(session.clj:47)
uest.session$do_with_current_user.invokeStatic(session.clj:48)
	at metabase.request.session$do_with_current_user.invoke(session.clj:30)
	at metabase.server.middleware.session$bind_current_user$fn__148180.invoke(session.clj:244)
	at metabase.server.middleware.session$wrap_current_user_info$fn__148172.invoke(session.clj:220)
	at metabase.analytics.sdk$embedding_mw$embedding_mw_fn__92972.invoke(sdk.clj:68)
	at metabase.server.middleware.session$wrap_session_key$fn__148138.invoke(session.clj:82)
	at metabase.server.middleware.auth$wrap_static_api_key$fn__150209.invoke(auth.clj:17)
	at ring.middleware.cookies$wrap_cookies$fn__151369.invoke(cookies.clj:200)
	at metabase.server.middleware.misc$add_content_type$fn__151065.invoke(misc.clj:28)
	at metabase.server.middleware.misc$disable_streaming_buffering$fn__151095.invoke(misc.clj:75)
	at ring.middleware.gzip$wrap_gzip$fn__151419.invoke(gzip.clj:86)
	at metabase.server.middleware.request_id$wrap_request_id$fn__151147.invoke(request_id.clj:10)
.server.middleware.misc$bind_request$fn__151098$fn__151099.invoke(misc.clj:87)
	at metabase.request.current$do_with_current_request.invokeStatic(current.clj:19)
	at metabase.request.current$do_with_current_request.invoke(current.clj:15)
	at metabase.server.middleware.misc$bind_request$fn__151098.invoke(misc.clj:86)
	at metabase.server.middleware.ssl$redirect_to_https_middleware$fn__151170.invoke(ssl.clj:42)
	at metabase.server.handler$fn__151530$dev_handler151529__151531$dev_handler_STAR___151548.invoke(handler.clj:109)
	at metabase.server.test_handler$fn__155583$make_test_handler155582__155584$fn__155585.invoke(test_handler.clj:19)
	at metabase.server.instance$async_proxy_handler$fn__150769.invoke(instance.clj:81)
	at metabase.server.instance.proxy$org.eclipse.jetty.ee9.servlet.ServletHandler$ff19274a.doHandle(Unknown Source)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.handle(ScopedHandler.java:127)

	at metabase.server.statistics_handler.proxy$org.eclipse.jetty.ee9.nested.HandlerWrapper$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.ee9.nested.HandlerWrapper.handle(HandlerWrapper.java:124)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.nextHandle(ScopedHandler.java:197)
	at org.eclipse.jetty.ee9.nested.ContextHandler.doHandle(ContextHandler.java:1045)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.nextScope(ScopedHandler.java:164)
	at org.eclipse.jetty.ee9.servlet.ServletHandler.doScope(ServletHandler.java:483)
	at metabase.server.instance.proxy$org.eclipse.jetty.ee9.servlet.ServletHandler$ff19274a.doScope(Unknown Source)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.nextScope(ScopedHandler.java:162)
	at org.eclipse.jetty.ee9.nested.ContextHandler.doScope(ContextHandler.java:966)
	at org.eclipse.jetty.ee9.nested.ScopedHandler.handle(ScopedHandler.java:125)
	at org.eclipse.jetty.ee9.nested.ContextHandler.handle(ContextHandler.java:1719)
uestDispatchable.dispatch(HttpChannel.java:1565)nel$Req
	at org.eclipse.jetty.ee9.nested.HttpChannel.dispatch(HttpChannel.java:724)
	at org.eclipse.jetty.ee9.nested.HttpChannel.handle(HttpChannel.java:512)
	at org.eclipse.jetty.ee9.nested.ContextHandler$CoreContextHandler$CoreToNestedHandler.handle(ContextHandler.java:3029)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1071)
	at org.eclipse.jetty.server.Server.handle(Server.java:182)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:678)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:416)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
.java:480) org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:443)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:201)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:311)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:981)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1211)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1166)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.sql.SQLException: Failed to create connection
	at com.clickhouse.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:104)
	at com.clickhouse.jdbc.Driver.connect(Driver.java:123)
er.connect(ClickHouseDriver.java:97)HouseDriv
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:683)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:191)
	at clojure.java.jdbc$get_driver_connection.invokeStatic(jdbc.clj:271)
	at clojure.java.jdbc$get_driver_connection.invoke(jdbc.clj:250)
	at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:411)
	at clojure.java.jdbc$get_connection.invoke(jdbc.clj:274)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1111)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1160)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invokeStatic(connection.clj:368)
	at metabase.driver.sql_jdbc.connection$can_connect_with_spec_QMARK_.invoke(connection.clj:365)
t metabase.driver.sql_jdbc.connection$can_connect_QMARK_$fn__137759.invoke(connection.clj:377)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection$fn__137737.invoke(connection.clj:353)
	at metabase.driver.sql_jdbc.connection.ssh_tunnel$do_with_ssh_tunnel.invokeStatic(ssh_tunnel.clj:152)
	at metabase.driver.sql_jdbc.connection.ssh_tunnel$do_with_ssh_tunnel.invoke(ssh_tunnel.clj:141)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invokeStatic(connection.clj:348)
	at metabase.driver.sql_jdbc.connection$do_with_connection_spec_for_testing_connection.invoke(connection.clj:344)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invokeStatic(connection.clj:376)
	at metabase.driver.sql_jdbc.connection$can_connect_QMARK_.invoke(connection.clj:372)
	at metabase.driver.clickhouse$eval200658$fn__200659.invoke(clickhouse.clj:172)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
ARK_$fn__89713.invoke(util.clj:135)_connect_with_details_QM
	at clojure.core$binding_conveyor_fn$fn__5844.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	... 1 more
Caused by: com.clickhouse.client.api.ClientException: Failed to get server info
	at com.clickhouse.client.api.Client.loadServerInfo(Client.java:204)
	at com.clickhouse.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:93)
	... 33 more
Caused by: com.clickhouse.client.api.ServerException: Code: 194. DB::Exception: default: Authentication failed: password is incorrect, or there is no user with such name. If you use Cl reset it in the configuration file. The password for default user is typically located at /etc/clickhouse-server/users.d/default-password.xml and deleting this file will reset the password. See also /etc/clickhouse-server/users.xml on the server where ClickHouse is installed. . (REQUIRED_PASSWORD) (version 25.1.3.23 (official build))
	at com.clickhouse.client.api.internal.HttpAPIClientHelper.readError(HttpAPIClientHelper.java:371)
	at com.clickhouse.client.api.internal.HttpAPIClientHelper.executeRequest(HttpAPIClientHelper.java:426)
	at com.clickhouse.client.api.Client.lambda$query$10(Client.java:1705)
	at com.clickhouse.client.api.Client.runAsyncOperation(Client.java:2134)
	at com.clickhouse.client.api.Client.query(Client.java:1748)
	at com.clickhouse.client.api.Client.query(Client.java:1634)
	at com.clickhouse.client.api.Client.loadServerInfo(Client.java:195)
	... 34 more
2025-06-27 21:22:41,623 DEBUG middleware.log :: PUT /api/database/383 400 290ms (1 DB calls) {:metabase-user-id 1}
{:message "Failed to create connection"}
```
</details>

contrast with an incorrect username in postgres:

<img width="622" alt="Image" src="https://github.com/user-attachments/assets/6530ff65-524e-4487-82d2-2509e476a6e6" />


### To Reproduce

connect with incorrect details

### Expected behavior

helpful feedback that the password is wrong

### Logs

_No response_

### Information about your Metabase installation

```JSON
masteer
```

### Severity

p2

### Additional context

_No response_

--- a/docs/developers-guide/driver-changelog.md
+++ b/docs/developers-guide/driver-changelog.md
@@ -17,6 +17,9 @@ title: Driver interface changelog
   metabase.driver/connection-details, metabase.driver/table-exists?, metabase.driver.sql/normalize-name,
   metabase.driver.sql/default-schema, and metabase.driver.sql/find-table to implement sql transforms.
 
+- `driver/humanize-connection-error-message` now takes a list of all messages in the exception chain,
+  instead of just the top-level message as a string.
+
 ## Metabase 0.56.3
 
 - Added the driver multi-method `driver/describe-database*` that drivers should now implement instead of `driver/describe-database`.

--- a/modules/drivers/clickhouse/src/metabase/driver/clickhouse.clj
+++ b/modules/drivers/clickhouse/src/metabase/driver/clickhouse.clj
@@ -188,6 +188,14 @@
   (when table-or-field-name
     (str/replace table-or-field-name #"-" "_")))
 
+(defmethod driver/humanize-connection-error-message :clickhouse
+  [_ messages]
+  (condp re-matches (str/join " -> " messages)
+    #".*AUTHENTICATION_FAILED.*"
+    :username-or-password-incorrect
+
+    (first messages)))
+
 ;;; ------------------------------------------ Connection Impersonation ------------------------------------------
 
 (defmethod driver/upload-type->database-type :clickhouse

--- a/modules/drivers/clickhouse/test/metabase/driver/clickhouse_test.clj
+++ b/modules/drivers/clickhouse/test/metabase/driver/clickhouse_test.clj
@@ -354,3 +354,9 @@
                               :target [:variable [:template-tag "category_id_2"]]
                               :value  "2"}]
                 :middleware {:format-rows? false}})))))))
+
+(deftest ^:parallel humanize-connection-error-message-test
+  (is (= "random message" (driver/humanize-connection-error-message :clickhouse ["random message"])))
+  (is (= :username-or-password-incorrect (driver/humanize-connection-error-message :clickhouse ["Failed to create connection"
+                                                                                                "Failed to get server info"
+                                                                                                "Code: 516. DB::Exception: asdf: Authentication failed: password is incorrect, or there is no user with such name. (AUTHENTICATION_FAILED) (version 25.7.4.11 (official build))"]))))

--- a/modules/drivers/mongo/src/metabase/driver/mongo.clj
+++ b/modules/drivers/mongo/src/metabase/driver/mongo.clj
@@ -60,36 +60,37 @@
 
 (defmethod driver/humanize-connection-error-message
   :mongo
-  [_ message]
-  (condp re-matches message
-    #"^Timed out after \d+ ms while waiting for a server .*$"
-    :cannot-connect-check-host-and-port
+  [_ messages]
+  (let [message (first messages)]
+    (condp re-matches message
+      #"^Timed out after \d+ ms while waiting for a server .*$"
+      :cannot-connect-check-host-and-port
 
-    #"^host and port should be specified in host:port format$"
-    :invalid-hostname
+      #"^host and port should be specified in host:port format$"
+      :invalid-hostname
 
-    #"^Password can not be null when the authentication mechanism is unspecified$"
-    :password-required
+      #"^Password can not be null when the authentication mechanism is unspecified$"
+      :password-required
 
-    #"^org.apache.sshd.common.SshException: No more authentication methods available$"
-    :ssh-tunnel-auth-fail
+      #"^org.apache.sshd.common.SshException: No more authentication methods available$"
+      :ssh-tunnel-auth-fail
 
-    #"^java.net.ConnectException: Connection refused$"
-    :ssh-tunnel-connection-fail
+      #"^java.net.ConnectException: Connection refused$"
+      :ssh-tunnel-connection-fail
 
-    #".*javax.net.ssl.SSLHandshakeException: PKIX path building failed.*"
-    :certificate-not-trusted
+      #".*javax.net.ssl.SSLHandshakeException: PKIX path building failed.*"
+      :certificate-not-trusted
 
-    #".*MongoSocketReadException: Prematurely reached end of stream.*"
-    :requires-ssl
+      #".*MongoSocketReadException: Prematurely reached end of stream.*"
+      :requires-ssl
 
-    #".* KeyFactory not available"
-    :unsupported-ssl-key-type
+      #".* KeyFactory not available"
+      :unsupported-ssl-key-type
 
-    #"java.security.InvalidKeyException: invalid key format"
-    :invalid-key-format
+      #"java.security.InvalidKeyException: invalid key format"
+      :invalid-key-format
 
-    message))
+      messages)))
 
 ;;; ### Syncing
 
@@ -175,12 +176,12 @@
 (def ^:private unwind-stages
   "Sequence of stages repeated in _search_ phase of [[describe-table-pipeline]]
     for [[describe-table-query-depth]] times.
-  
+
     Each repetion $unwinds documents having `val` of type \"object\", so those are __swapped__ for sequence
     of their children.
-  
+
     Documents with non-object val are left untouched.
-  
+
     Each document that is processed has path from parent stored in `path`. `indices` represent indices of keys
     in the `path` in parent objects as per $objectToArray."
   [{"$addFields" {"kvs" {"$cond" [{"$eq" [{"$type" "$val"} "object"]} {"$objectToArray" "$val"} nil]}}}
@@ -359,7 +360,7 @@
   first.
 
   The resulting structure will hold at most [[driver.settings/sync-leaf-fields-limit]] leaf fields. That translates
-  to at most [[driver.settings/sync-leaf-fields-limit]] * [[describe-table-query-depth]]. That is 7K fields at the 
+  to at most [[driver.settings/sync-leaf-fields-limit]] * [[describe-table-query-depth]]. That is 7K fields at the
   time of writing hence safe to reside in memory for further operation."
   [dbfields]
   (-> dbfields dbfields->ftree* ftree-reconcile-nodes))

--- a/modules/drivers/mongo/src/metabase/driver/mongo.clj
+++ b/modules/drivers/mongo/src/metabase/driver/mongo.clj
@@ -90,7 +90,7 @@
       #"java.security.InvalidKeyException: invalid key format"
       :invalid-key-format
 
-      messages)))
+      message)))
 
 ;;; ### Syncing
 

--- a/modules/drivers/oracle/src/metabase/driver/oracle.clj
+++ b/modules/drivers/oracle/src/metabase/driver/oracle.clj
@@ -571,12 +571,13 @@
   (sql.qp/->honeysql driver [::sql.qp/cast expr "varchar2(256)"]))
 
 (defmethod driver/humanize-connection-error-message :oracle
-  [_ message]
+  [_ messages]
   ;; if the connection error message is caused by the assertion above checking whether sid or service-name is set,
   ;; return a slightly nicer looking version. Otherwise just return message as-is
-  (if (str/includes? message "(or sid service-name)")
-    "You must specify the SID and/or the Service Name."
-    message))
+  (let [message (first messages)]
+    (if (str/includes? message "(or sid service-name)")
+      "You must specify the SID and/or the Service Name."
+      message)))
 
 (defn- remove-rownum-column
   "Remove the `:__rownum__` column from results, if present."

--- a/modules/drivers/oracle/test/metabase/driver/oracle_test.clj
+++ b/modules/drivers/oracle/test/metabase/driver/oracle_test.clj
@@ -104,7 +104,7 @@
            (try (sql-jdbc.conn/connection-details->spec :oracle {:host "localhost"
                                                                  :port 1521})
                 (catch Throwable e
-                  (driver/humanize-connection-error-message :oracle (.getMessage e))))))))
+                  (driver/humanize-connection-error-message :oracle (u/all-ex-messages e))))))))
 
 (deftest connection-properties-test
   (testing "Connection properties should be returned properly (including transformation of secret types)"

--- a/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
+++ b/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
@@ -68,14 +68,15 @@
   (defmethod driver/database-supports? [:snowflake feature] [_driver _feature _db] supported?))
 
 (defmethod driver/humanize-connection-error-message :snowflake
-  [_ message]
-  (log/spy :error (type message))
-  (condp re-matches message
-    #"(?s).*Object does not exist.*$"
-    :database-name-incorrect
+  [_ messages]
+  (let [message (first messages)]
+    (log/spy :error (type message))
+    (condp re-matches message
+      #"(?s).*Object does not exist.*$"
+      :database-name-incorrect
 
-    ; default - the Snowflake errors have a \n in them
-    message))
+      ; default - the Snowflake errors have a \n in them
+      message)))
 
 (defmethod driver/db-start-of-week :snowflake
   [_]

--- a/src/metabase/driver.clj
+++ b/src/metabase/driver.clj
@@ -842,13 +842,14 @@
   as keywords whenever possible. This provides for both unified error messages and categories which let us point
   users to the erroneous input fields.
   Error messages can also be strings, or localized strings, as returned by [[metabase.util.i18n/trs]] and
-  `metabase.util.i18n/tru`."
-  {:added "0.32.0" :arglists '([this message])}
+  `metabase.util.i18n/tru`.
+  Passed a collection of all non-nil exception messages that were thrown during connection attempt."
+  {:added "0.32.0" :arglists '([this messages])}
   dispatch-on-initialized-driver
   :hierarchy #'hierarchy)
 
-(defmethod humanize-connection-error-message ::driver [_ message]
-  message)
+(defmethod humanize-connection-error-message ::driver [_ messages]
+  (first messages))
 
 (defmulti mbql->native
   "Transpile an MBQL query into the appropriate native query form. `query` will match the schema for an MBQL query in

--- a/src/metabase/driver/h2.clj
+++ b/src/metabase/driver/h2.clj
@@ -307,18 +307,19 @@
     (dateadd unit amount hsql-form)))
 
 (defmethod driver/humanize-connection-error-message :h2
-  [_ message]
-  (condp re-matches message
-    #"^A file path that is implicitly relative to the current working directory is not allowed in the database URL .*$"
-    :implicitly-relative-db-file-path
+  [_ messages]
+  (let [message (first messages)]
+    (condp re-matches message
+      #"^A file path that is implicitly relative to the current working directory is not allowed in the database URL .*$"
+      :implicitly-relative-db-file-path
 
-    #"^Database .* not found, .*$"
-    :db-file-not-found
+      #"^Database .* not found, .*$"
+      :db-file-not-found
 
-    #"^Wrong user name or password .*$"
-    :username-or-password-incorrect
+      #"^Wrong user name or password .*$"
+      :username-or-password-incorrect
 
-    message))
+      message)))
 
 (defmethod driver/db-default-timezone :h2
   [_driver _database]

--- a/src/metabase/driver/mysql.clj
+++ b/src/metabase/driver/mysql.clj
@@ -244,22 +244,23 @@
   (h2x/current-datetime-honeysql-form driver))
 
 (defmethod driver/humanize-connection-error-message :mysql
-  [_ message]
-  (condp re-matches message
-    #"^Communications link failure\s+The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.$"
-    :cannot-connect-check-host-and-port
+  [_ messages]
+  (let [message (first messages)]
+    (condp re-matches message
+      #"^Communications link failure\s+The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.$"
+      :cannot-connect-check-host-and-port
 
-    #"^Unknown database .*$"
-    :database-name-incorrect
+      #"^Unknown database .*$"
+      :database-name-incorrect
 
-    #"Access denied for user.*$"
-    :username-or-password-incorrect
+      #"Access denied for user.*$"
+      :username-or-password-incorrect
 
-    #"Must specify port after ':' in connection string"
-    :invalid-hostname
+      #"Must specify port after ':' in connection string"
+      :invalid-hostname
 
-    ;; else
-    message))
+      ;; else
+      message)))
 
 #_{:clj-kondo/ignore [:deprecated-var]}
 (defmethod sql-jdbc.sync/db-default-timezone :mysql

--- a/src/metabase/driver/postgres.clj
+++ b/src/metabase/driver/postgres.clj
@@ -110,28 +110,29 @@
 (defmethod driver/display-name :postgres [_] "PostgreSQL")
 
 (defmethod driver/humanize-connection-error-message :postgres
-  [_ message]
-  (condp re-matches message
-    #"^FATAL: database \".*\" does not exist$"
-    :database-name-incorrect
+  [_ messages]
+  (let [message (first messages)]
+    (condp re-matches message
+      #"^FATAL: database \".*\" does not exist$"
+      :database-name-incorrect
 
-    #"^No suitable driver found for.*$"
-    :invalid-hostname
+      #"^No suitable driver found for.*$"
+      :invalid-hostname
 
-    #"^Connection refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.$"
-    :cannot-connect-check-host-and-port
+      #"^Connection refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.$"
+      :cannot-connect-check-host-and-port
 
-    #"^FATAL: role \".*\" does not exist$"
-    :username-incorrect
+      #"^FATAL: role \".*\" does not exist$"
+      :username-incorrect
 
-    #"^FATAL: password authentication failed for user.*$"
-    :password-incorrect
+      #"^FATAL: password authentication failed for user.*$"
+      :password-incorrect
 
-    #"^FATAL: .*$" ; all other FATAL messages: strip off the 'FATAL' part, capitalize, and add a period
-    (let [[_ message] (re-matches #"^FATAL: (.*$)" message)]
-      (str (str/capitalize message) \.))
+      #"^FATAL: .*$" ; all other FATAL messages: strip off the 'FATAL' part, capitalize, and add a period
+      (let [[_ message] (re-matches #"^FATAL: (.*$)" message)]
+        (str (str/capitalize message) \.))
 
-    message))
+      message)))
 
 (defmethod driver/db-default-timezone :postgres
   [driver database]

--- a/src/metabase/driver/util.clj
+++ b/src/metabase/driver/util.clj
@@ -138,7 +138,7 @@
       ;; first
       (catch Throwable e
         (log/error e "Failed to connect to Database")
-        (throw (if-let [humanized-message (some->> (.getMessage e)
+        (throw (if-let [humanized-message (some->> (u/all-ex-messages e)
                                                    (driver/humanize-connection-error-message driver))]
                  (let [error-data (cond
                                     (keyword? humanized-message)

--- a/src/metabase/util.cljc
+++ b/src/metabase/util.cljc
@@ -60,6 +60,7 @@
 
 #?(:clj (p/import-vars [u.jvm
                         all-ex-data
+                        all-ex-messages
                         auto-retry
                         string-to-bytes
                         bytes-to-string

--- a/src/metabase/util/jvm.clj
+++ b/src/metabase/util/jvm.clj
@@ -149,6 +149,17 @@
    nil
    (full-exception-chain e)))
 
+(defn all-ex-messages
+  "Returns a list of all non-nil messages in an exception, starting from the outermost and working inward.
+  If no messages are found, returns nil."
+  [^Throwable e]
+  (loop [ex e
+         messages []]
+    (if ex
+      (recur (.getCause ex)
+             (conj messages (.getMessage ex)))
+      (seq (remove nil? messages)))))
+
 (defn do-with-auto-retries
   "Execute `f`, a function that takes no arguments, and return the results.
    If `f` fails with an exception, retry `f` up to `num-retries` times until it succeeds.

--- a/src/metabase/warehouses/models/database.clj
+++ b/src/metabase/warehouses/models/database.clj
@@ -213,7 +213,7 @@
            (analytics/inc! :metabase-database/status {:driver engine :healthy true})
 
            (catch Throwable e
-             (let [humanized-message (some->> (.getMessage e)
+             (let [humanized-message (some->> (u/all-ex-messages e)
                                               (driver/humanize-connection-error-message driver))
                    reason (if (keyword? humanized-message) "user-input" "exception")]
                (log/error e (u/format-color :red "Health check: failure with error %s {:id %d :reason %s :message %s}"

--- a/test/metabase/util/jvm_test.clj
+++ b/test/metabase/util/jvm_test.clj
@@ -106,3 +106,19 @@
     "¥200."         200.0M
     "$.05"          0.05M
     "0.05"          0.05M))
+
+(deftest ^:parallel all-ex-messages-test
+  (testing "nil exception"
+    (is (nil? (u/all-ex-messages nil))))
+  (testing "one level exception"
+    (is (= ["test string"] (u/all-ex-messages (RuntimeException. "test string"))))
+    (is (nil? (u/all-ex-messages (Exception. nil nil)))))
+  (testing "chained exceptions"
+    (is (= ["test string 1" "test string 2"]
+           (u/all-ex-messages (RuntimeException. "test string 1" (RuntimeException. "test string 2")))))
+    (is (= ["test string 1" "test string 2"]
+           (u/all-ex-messages (RuntimeException. "test string 1" (RuntimeException. nil (RuntimeException. "test string 2"))))))
+    (is (= ["test string 2"]
+           (u/all-ex-messages (RuntimeException. nil (RuntimeException. "test string 2")))))
+    (is (nil?
+         (u/all-ex-messages (RuntimeException. nil (RuntimeException. nil nil)))))))

